### ISSUE TITLE
Coroutine tasks cannot be removed if they're awaiting.

### ISSUE THREAD

**Issue by LittleToonCat** [panda3d/panda3d#911]
Trying to remove a coroutine task (through `task.remove()` or `taskMgr.remove(task)`) while it's awaiting for another task to finish does not seem to kill the running task, it keeps on awaiting instead.  It'll be a pain to someone if they're planning to stop a coroutine on a `while True` loop doing some game logic.

It seems like the engine does not currently know how to remove the task if it's under the `S_awaiting` state.

```
(when starting the coroutine task)
:task(debug): Adding PythonTask testTask
:task(debug): Adding PythonTask continuousTask
:task(debug): PythonTask testTask is now awaiting <PythonTask continuousTask>.

(when attempting to stop the same task)
:task(debug): Removing PythonTask testTask
:task(debug):   (unable to remove PythonTask testTask)
```

Here's the code that reproduces the bug:
```python
from panda3d.core import *
from direct.showbase.ShowBase import ShowBase

loadPrcFileData('', 'notify-level-task debug')

def continuousTask(task):
    # Sample task that goes on forever.
    return task.cont

async def testTask():
    # Sample coroutine that awaits forever.
    await taskMgr.add(continuousTask)

class MyApp(ShowBase):
    def __init__(self):
        ShowBase.__init__(self)
        # Coroutine task
        self.task = taskMgr.add(testTask())

        self.accept('escape', self.__stopTask)

    def __stopTask(self):
        print(self.task.state) # Prints out 6 (S_awaiting)
        print(self.task.remove()) # This is supposed to stop the task, but it returns False instead. 

app = MyApp()
app.run()
```

**Comment by rdb** [comment]
Hmm, interesting.  What I'm wondering, though, is whether it should throw an exception into the task (so that any `finally` blocks get a chance to run, and clean up any resources), or whether it should just remove the task at the `await` block.  I worry that if we just stop the task right in the middle of it, resources held by the task (like open files) may not be released properly.

**Comment by LittleToonCat** [comment]
I think it pretty much varies of what you're doing with that coroutine.  I understand that coroutines are mainly used for loading resources asynchronously, But what I'm mainly using them for is to mimic the old retro adventure games. (calling a coroutine for dialog which awaits till a character has finished speaking, starting background tasks, etc.)  So tasks and coroutines would handle most of the game logic (Having coroutines is great, because I don't have to write callbacks practically everywhere).

Back to your question.  I think having it throw an exception would be very nice, that way I could clean up if the certain coroutine asks for it.  But then again, it would be also nice if it could remove the tasks that its currently awaiting for too, one less task I have to cleanup.  I guess this can be handled either (or both) way, but this up to you really.  I can write `finally` blocks if I have too.

**Comment by rdb** [comment]
Perhaps we can have separate methods, with `remove()` only removing it after the coroutine is done, `cancel()` to throw an exception into the coroutine, and `kill()` to immediately stop the coroutine without cleaning up resources.

**Comment by LittleToonCat** [comment]
That sounds good. :+1:

**Comment by LittleToonCat** [comment]
Oh by the way.  You don't have to do this, but it'll be also be nice if such tasks can be readded to the TaskManager just like normal tasks can, starting exactly where they awaiting from, something like a `pause()` functionality  It'll be helpful if someone is planning on adding a pause functionality into their games, which mainly involves halting the coroutines, and starting back up again after unpausing.

**Comment by rdb** [comment]
> it'll be also be nice if such tasks can be readded to the TaskManager just like normal tasks can

I don't think we can do this anymore because I had made the design decision early on to make tasks inherit from our future class (to match the design of Python's asyncio), and futures (by their nature) can transition from the "running" to the "done" state only once in their lifetime.  So once a task is "done" (which includes "cancelled"), it cannot go back to the "running" state.

However, there *may* be a way to just suspend the task chain that these tasks are on entirely, so that you could prevent them from continuing past their current `await` point without having to cancel them entirely.  That's probably a better way to implement pausing than cancelling your coroutines.

Alternatively, a task receiving a CancelledError doesn't *have* to actually cancel itself; if it receives one it could simply await some other AsyncFuture that is flagged when the game is resumed, and then go back to what it was doing.  But it would probably involve a lot of boilerplate to do that everywhere.

**Comment by rdb** [comment]
This is now implemented; calling `remove()` on an awaiting coroutine task will still fail, but calling `cancel()` will make the `await` trigger a CancelledError, which (if uncaught) will cause the task to be cancelled.  This will cause any resources to be cleaned up properly.  It can be caught explicitly to customise the behaviour.

I would prefer not to implement `kill()` unless someone indicates a strong need for it.
